// libp2p/dist/src/insecure/pb/proto.js
/* eslint-disable import/export */
/* eslint-disable @typescript-eslint/no-namespace */
import { encodeMessage, decodeMessage, message, bytes, enumeration } from 'protons-runtime';
export var Exchange;
(function (Exchange) {
    Exchange.codec = () => {
        return message({
            1: { name: 'id', codec: bytes, optional: true },
            2: { name: 'pubkey', codec: PublicKey.codec(), optional: true }
        });
    };
    Exchange.encode = (obj) => {
        return encodeMessage(obj, Exchange.codec());
    };
    Exchange.decode = (buf) => {
        return decodeMessage(buf, Exchange.codec());
    };
})(Exchange || (Exchange = {}));
export var KeyType;
(function (KeyType) {
    KeyType["RSA"] = "RSA";
    KeyType["Ed25519"] = "Ed25519";
    KeyType["Secp256k1"] = "Secp256k1";
    KeyType["ECDSA"] = "ECDSA";
})(KeyType || (KeyType = {}));
var __KeyTypeValues;
(function (__KeyTypeValues) {
    __KeyTypeValues[__KeyTypeValues["RSA"] = 0] = "RSA";
    __KeyTypeValues[__KeyTypeValues["Ed25519"] = 1] = "Ed25519";
    __KeyTypeValues[__KeyTypeValues["Secp256k1"] = 2] = "Secp256k1";
    __KeyTypeValues[__KeyTypeValues["ECDSA"] = 3] = "ECDSA";
})(__KeyTypeValues || (__KeyTypeValues = {}));
(function (KeyType) {
    KeyType.codec = () => {
        return enumeration(__KeyTypeValues);
    };
})(KeyType || (KeyType = {}));
export var PublicKey;
(function (PublicKey) {
    PublicKey.codec = () => {
        return message({
            1: { name: 'Type', codec: KeyType.codec() },
            2: { name: 'Data', codec: bytes }
        });
    };
    PublicKey.encode = (obj) => {
        return encodeMessage(obj, PublicKey.codec());
    };
    PublicKey.decode = (buf) => {
        return decodeMessage(buf, PublicKey.codec());
    };
})(PublicKey || (PublicKey = {}));
